x8.md

```diff
@@ -41,160 +45,29 @@ As you can see, the message is an array and each index value has its meaning:
-
-The implementation of this protocol is mainly done in https://github.com/status-im/status-react and https://github.com/status-im/status-go repositories.
-
-The goal of this protocol is to allow people running Ethereum nodes with Whisper service enabled to exchange messages that are end-to-end encrypted in a way that guarantees [darkness to some extent](https://github.com/ethereum/go-ethereum/wiki/Achieving-Darkness).
-
-It's important to notice that messages [are not limited to be text messages](#content-types) only. They can also have special meaning depending on the client implementation. For example, in the current implementation, there are message which informs about Eth requests.
-
-This protocol consist of three components:
-* payload
-* Whisper adapter
-* offline messaging.
-
-[The payload section](#payload) describes how the messages are encoded and decoded and what each fields of a message means. This is required to properly interpret messages by the client.
-
-Whisper adapter specifies interaction with the Whisper service with regards to keys management, configuration and attaching metadata required to properly forward and process messages.
-
-Offline messaging describes how the protocol handles delivering messages when one or more participants are offline and the messages expire in the Whisper network.
-
-The protocol does not specify additional things like peers discovery, running Whisper nodes, underlying p2p protocols etc.
-
-# Payload
-
-Payload is an array of bytes exchanged between peers. In the case of this protocol, the payload is encoded using [transit format](https://github.com/cognitect/transit-format). It is a text-based format so even encoded version is easily readable by humans.
-
-Payload contains the following fields:
-1. text `string`
-2. content type `enum { text/plain }` (more in [Content types](#content-types))
-3. message type `enum { public-group-user-message, user-message, group-user-message }` (more in [Message types](#message-types))
-4. clock `int64`
-5. timestamp `int64`
-6. content `struct { chat-id string, text string }`
-
-Example of a valid encoded payload:
-
-```
-["~#c4",["abc123","text/plain","~:public-group-user-message",154593077368201,1545930773682,["^ ","~:chat-id","testing-adamb","~:text","abc123"]]]
-```
+## Future encoding format
 
-As you can see, the message is an array and each index value has its meaning:
-* 0: `c4` is a decoder handler identification for the current payload format. Identifications allow to register handlers for many different types of payload
-* 1: array which items correspond to the described payload fields above
+Although, Transit format suits for most of the Status Protocol needs, it has drawbacks:
+* it is not widely popular and implementations for less popular languages might not be complete or exist at all
+* upgradability is a concern while, for example, protocol buffers are easily extensible.
```

Comment:
is this true or just about how we use it?

Reply:
not sure how true this statement is, there might be some truth in it, but the differences between transit and protobuf are not that much, in terms of upgradability.

I know this sound trivial, but:

Very simply, protobuf is a map, Transit is an array, and that's really all there is to it.
You add new fields to a map by adding a key, you add fields in an array by appending a member.
You add a field to both, you don't break compatibility, you remove a field from either, you might break compatibility.

Is a map more upgradable then an array? Sure maps are more suited for these kind of things,
 and arrays make certain things harder (removing fields,for example, where you would want to keep positional order), but if you are removing fields, you likely are already breaking compatibility, so since adding is essentially identical and removing/changing is the only way to break compatibility, I would say that in terms of upgradability they are equivalent.

(there are are differences of course, one is typed the other isn't etc, but do not concern upgradability)

Reply:
Actually transit could have been using a map has well, the thing is the protocol was designed with upgradability in mind not forever backward and forward compatibility, the idea was not to extend the records but to use a new version and publish intermediary build that would support both.

So this is really not something that is desirable, just something we have to deal with for as long as we don't fix it by breaking compatibility once

Reply:
> Very simply, protobuf is a map, Transit is an array, and that's really all there is to it.

In map the order is irrelevant while it is in an array. So, in transit I need to be careful to add the filed at the end, correct? In protobuf I specified field index.

Also, I was looking into that, and I couldn't find anything in the Transit repo that talks about upgradability while protobuf has a section dedicated to that: https://developers.google.com/protocol-buffers/docs/proto#updating.

So, are you for or against keeping this point?

The first statement is of course true, there is no good lib for Go, for example :D

Reply:
@adambabik correct, it's positional

for upgradability of transit, I believe is just because it's much simpler than protobuf (no types, no default values etc). Knowing the authors their thought is to discourage breaking changes, therefore choosing an array, which makes it awkward to remove fields, an encourage people to either append new fields, or use a completely different message type "tag".

In general I would word it probably like @yenda , the fault is not in transit, but mainly our own protocol was not thought out to be extended in the way we think of now.

Reply:
Again, transit could use a map as well, right now it uses a record, which is a kinda typed array, the reason was upgradability, but because we traded it for extensibility, it is completely useless, and we lost track of versions

Reply:
oh I stand corrected about arrays in transit, I did not know that, thanks @yenda